eng/targets/ResolveReferences.targets

```diff
@@ -148,6 +148,68 @@
            Text="Could not resolve this reference. Could not locate the package or project for &quot;%(Reference.Identity)&quot;" />
   </Target>
 
+  <!--
+    Remove @(ResolvedCompileFileDefinitions) and @(RuntimeCopyLocalItems) items for patched assemblies between item
+    generation and use. Instead, ensure @(ProjectReference) items include the patched projects. This covers indirect
+    references and improves test coverage; @(ProjectReference) items were set correctly for direct references. Changes
+    are necessary only in servicing builds, in the release/2.1 branch, and when building non-implementation projects.
+    This is imperfect because it covers only indirect references to project reference providers i.e. most C#
+    implementation projects and does not work for legacy samples or other non-.NET SDK projects.
+  -->
+  <Target Name="_TestPatchedAssemblies"
+      Condition=" $(UseProjectReferences) AND '$(PackagesInPatch)' != '' AND
+          (@(ResolvedCompileFileDefinitions->Count()) != 0 OR @(RuntimeCopyLocalItems->Count()) != 0) "
+      AfterTargets="ResolvePackageAssets"
+      BeforeTargets="AssignProjectConfiguration;
+          GenerateBuildDependencyFile;
+          GeneratePublishDependencyFile;
+          ILLink;
+          ResolveLockFileCopyLocalFiles;
+          ResolveLockFileReferences"
+      DependsOnTargets="ResolvePackageAssets">
+    <!-- Find projects patched in this build that need to be added as references. -->
+    <ItemGroup>
+      <_ProjectsInPatch Remove="@(_ProjectsInPatch)" />
+      <_ProjectsInPatch Include="@(ProjectReferenceProvider)"
+          Condition=" $(PackagesInPatch.Contains(' %(Identity);')) " />
+      <!-- dotnet-razorpagegenerator project is unique in overriding its package ID. -->
+      <_ProjectsInPatch Include="@(ProjectReferenceProvider->WithMetadataValue('Identity', 'dotnet-razorpagegenerator'))"
+          Condition=" $(PackagesInPatch.Contains(' RazorPageGenerator;')) " />
+
+      <!-- Ignore unreferenced projects/assemblies. Project reference providers have matching assembly and project names. -->
+      <_ProjectsToAdd Remove="@(_ProjectsToAdd)" />
+      <_ProjectsToAdd Include="@(_ProjectsInPatch)"
+          Condition=" @(RuntimeCopyLocalItems->WithMetadataValue('Filename', '%(_ProjectsToAdd.Identity)')->Count()) != 0 OR
+               @(ResolvedCompileFileDefinitions->WithMetadataValue('Filename', '%(_ProjectsToAdd.Identity)')->Count()) != 0 " />
```

Comment:
@rainersigwald I'm not able to get these item functions to do what I expect. Suggestions much appreciated❕ Note the target is _not_ batched.

Reply:
Idea is to check through `@(RuntimeCopyLocalItems)` and `@(ResolvedCompileFileDefinitions)` for items with filenames matching item specs in `@(_ProjectsToAdd)`.

I didn't do intersection using `<JoinItems>` or like https://github.com/dotnet/aspnetcore/blob/main/eng/targets/ResolveReferences.targets#L96-L125 because the requirements _should_ be simpler here. We don't need any metadata from either  `@(RuntimeCopyLocalItems)` or `@(ResolvedCompileFileDefinitions)` items.

I could create properties for filenames of the two item groups then put e.g. `$(RuntimeCopyLocalItems.Contains(' %(Identity);'))` in here. But, that seems like an unnecessary indirection. Hoping there's a better approach&hellip;